src/java/org/apache/cassandra/index/sai/disk/v1/postings/MergePostingList.java

```diff
@@ -98,18 +113,40 @@ else if (next == lastRowId)
     @Override
     public long advance(long targetRowID) throws IOException
     {
-        temp.clear();
-
-        for (var peekable : postingLists)
+        // clean out obsolete child lists, and remember the smallest row seen in case
+        // we can use it for the fast path
+        long nextRowId = PostingList.END_OF_STREAM;
+        PostingList nextPostingList = null;
+        var it = postingLists.iterator();
+        while (it.hasNext())
         {
+            var peekable = it.next();
             peekable.advanceWithoutConsuming(targetRowID);
-            if (peekable.peek() != PostingList.END_OF_STREAM)
-                temp.add(peekable);
+            long peeked = peekable.peek();
+            // clean out obsolete child lists
+            if (peeked == PostingList.END_OF_STREAM)
+                it.remove();
+
+            if (lastRowId <= peeked && peeked < nextRowId)
+            {
+                nextRowId = peeked;
+                nextPostingList = peekable;
```

Comment:
If we overwrite this multiple times, are we hitting the same problem below where we have an `an arbitrary number of our child posting lists may be pointing to`?

Reply:
Actually, it doesn't matter if we overwrite `nextPostingList`. It would only matter if we have the same `nextRowId` multiple times.

Reply:
Yes, if there is a tie and multiple PL have the same next row id, it doesn't matter which one we take from.